core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java

```diff
@@ -644,53 +693,166 @@ public void assertExpectedUndoMoveScore(Move<Solution_> move, Score_ beforeMoveS
         if (!undoScore.equals(beforeMoveScore)) {
             logger.trace("        Corruption detected. Diagnosing...");
 
-            ShadowVariablesAssert afterUndoSnapshot =
-                    ShadowVariablesAssert.takeSnapshot(getSolutionDescriptor(), workingSolution);
+            AllVariablesAssert<Solution_> afterUndoSnapshot =
+                    AllVariablesAssert.takeSnapshot(getSolutionDescriptor(), workingSolution);
+            Solution_ afterUndoSolution = cloneSolution(workingSolution);
             // Precondition: assert that there are probably no corrupted constraints
             assertWorkingScoreFromScratch(undoScore, undoMoveText);
             // Precondition: assert that shadow variables aren't stale after doing the undoMove
             assertShadowVariablesAreNotStale(undoScore, undoMoveText);
 
-            String differentShadowVariables = buildShadowVariableDiff(afterUndoSnapshot);
+            String differentVariables = buildVariableDiff(afterUndoSnapshot);
             String scoreDifference = undoScore.subtract(beforeMoveScore).toShortString();
 
-            throw new IllegalStateException("UndoMove corruption (" + scoreDifference
+            throw new UndoScoreCorruptionException("UndoMove corruption (" + scoreDifference
                     + "): the beforeMoveScore (" + beforeMoveScore + ") is not the undoScore (" + undoScore
                     + ") which is the uncorruptedScore (" + undoScore + ") of the workingSolution.\n"
-                    + differentShadowVariables
+                    + differentVariables
                     + "  1) Enable EnvironmentMode " + EnvironmentMode.FULL_ASSERT
                     + " (if you haven't already) to fail-faster in case there's a score corruption or variable listener corruption.\n"
                     + "  2) Check the Move.createUndoMove(...) method of the moveClass (" + move.getClass() + ")."
                     + " The move (" + move + ") might have a corrupted undoMove (" + undoMoveText + ").\n"
                     + "  3) Check your custom " + VariableListener.class.getSimpleName() + "s (if you have any)"
                     + " for shadow variables that are used by score constraints that could cause"
-                    + " the scoreDifference (" + scoreDifference + ").");
+                    + " the scoreDifference (" + scoreDifference + ").",
```

Comment:
IMO we can also do a lot here to make the string much more readable, raw strings were made for this.